### PR TITLE
Fix bugs in map composition operators when `parSafe=true`

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
Fix bugs in map composition operators when `parSafe=true` (#17542)

The pull #17413 fixed several bugs in the composition operators for
the `set` type. This PR fixes similar bugs in the `map` type.

Add a test which covers every composition operator, compound
assignment operator, and every combination of `parSafe` value over
the `map` type. Use the test to fix several bugs in the `map`
type, including:

  - Loosen the typing of composition operators for the `map` type
    so that the `parSafe` value of the LHS and RHS maps do not
    need to match
  - Ditto for the assignment operator
  - Replace several instances of direct indexing with calls to
    `add()` or `getValue()` as appropriate
  - Add a "no doc" cast operator to make the compiler happy
  - In `init=`, respect a partially instantiated left-hand-side (e.g. `map(int)`)

Also address a performance TODO for the `&` operator.

Reviewed by @daviditen. Thanks!

TESTING:

- [x] `ALL` on `linux64` when `COMM=none`
- [x] `ALL` on `linux64` when `COMM=gasnet`

Signed-off-by: David Longnecker <dlongnecke-cray@users.noreply.github.com>

### CODE REVIEW

**Comment by dlongnecke-cray** [comment]
Hi @daviditen would you be willing to review these changes to the `map` operators?

File: modules/standard/Map.chpl [hunk 1]
```diff
@@ -704,17 +722,31 @@ module Map {
...
+    // should also be changed accordingly.
+    if k1 != k2 then
+      compilerError('Cannot cast to map with different ',
+                    'key type: ', k2:string);
+
+    if v1 != v2 then
+      compilerError('Cannot cast to map with different ',
+                    'value type: ', v2:string);
+
+    var result: map(k1, v1, p2) = x;
```
Comment (daviditen) [inline modules/standard/Map.chpl:747]: Can this just have `var result: t = x`? Or if you want to spell it out I'd prefer to use `k2`, `v2`, and `p2`.

File: modules/standard/Map.chpl [hunk 2]
```diff
@@ -759,108 +791,108 @@ module Map {
...
   }
 
   /*
     Sets the left-hand side map to contain the keys and values in either
     a or b.
    */
-  operator map.+=(ref a: map(?keyType, ?valueType, ?parSafe),
-                  b: map(keyType, valueType, parSafe)) {
+  operator map.+=(ref a: map(?keyType, ?valueType, ?),
+                      b: map(keyType, valueType, ?)) {
```
Comment (daviditen) [inline modules/standard/Map.chpl:804]: I'd prefer having a single indentation strategy when splitting lines. Other places line the second line up right after the opening parenthesis.

File: modules/standard/Map.chpl [hunk 3]
```diff
@@ -759,108 +791,108 @@ module Map {
...
     return newMap;
   }
 
   /* Sets the left-hand side map to contain the keys and values in either
      a or b.
    */
-  operator map.|=(ref a: map(?keyType, ?valueType, ?parSafe),
-                  b: map(keyType, valueType, parSafe)) {
+  operator map.|=(ref a: map(?keyType, ?valueType, ?),
+                      b: map(keyType, valueType, ?)) {
```
Comment (daviditen) [inline modules/standard/Map.chpl:822]: Indentation.

File: modules/standard/Map.chpl [hunk 4]
```diff
@@ -759,108 +791,108 @@ module Map {
...
+                  b: map(keyType, valueType, ?)) {
     a = a & b;
   }
 
   /* Returns a new map containing the keys that are only in a, but not b. */
-  operator map.-(a: map(?keyType, ?valueType, ?parSafe),
-                 b: map(keyType, valueType, parSafe)) {
-    var newMap = new map(keyType, valueType, parSafe);
+  operator map.-(a: map(?keyType, ?valueType, ?),
+             b: map(keyType, valueType, ?)) {
```
Comment (daviditen) [inline modules/standard/Map.chpl:848]: Indentation.

File: modules/standard/Map.chpl [hunk 5]
```diff
@@ -759,108 +791,108 @@ module Map {
...
 
     return newMap;
   }
 
   /* Sets the left-hand side map to contain the keys that are in the
      left-hand map, but not the right-hand map. */
-  operator map.-=(ref a: map(?keyType, ?valueType, ?parSafe),
-                  b: map(keyType, valueType, parSafe)) {
+  operator map.-=(ref a: map(?keyType, ?valueType, ?),
+                      b: map(keyType, valueType, ?)) {
```
Comment (daviditen) [inline modules/standard/Map.chpl:862]: Indentation.

File: modules/standard/Map.chpl [hunk 6]
```diff
@@ -759,108 +791,108 @@ module Map {
...
+      if !a.contains(k) then newMap.add(k, b.getValue(k));
     return newMap;
   }
 
   /* Sets the left-hand side map to contain the keys that are in either the
      left-hand map or the right-hand map, but not both. */
-  operator map.^=(ref a: map(?keyType, ?valueType, ?parSafe),
-                  b: map(keyType, valueType, parSafe)) {
+  operator map.^=(ref a: map(?keyType, ?valueType, ?),
+                      b: map(keyType, valueType, ?)) {
```
Comment (daviditen) [inline modules/standard/Map.chpl:892]: Indentation.

**Review by daviditen — APPROVED** [review]
Looks good other than minor nits. Thanks!

**Pull request merged**